### ISSUE TITLE
Server DB Normalization on user table

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#2842]
Wanted to open a thread to developers for a heads up on intended server-side DB updates.

Intended update: normalize the user table, introduce a synthetic PK column.

Why?
The concept of a 'user' is going to be of even more importance soon and we'll likely need a few tables with FK relationships to users. I have the 'match challenge' feature on my radar.

As long as the user table PK column is the username, and is a string, FK relationships become very complicated to manage, if not impossible if we still want to allow users to change their name.

Some basic DB normalization solves this problem cleanly.

**Comment by ssoloff** [comment]
@DanVanAtta Note that once #2828 is merged, I'm going to submit a follow-up PR rather quickly to do the same thing for the mute tables, as requested by @prastle.  I'd like to get that in before any big schema changes, if possible.  It shouldn't take more than 24 hours to submit the new PR once #2828 is merged due to the similarity between the two.  So hopefully that won't hold you up any.

**Comment by DanVanAtta** [comment]
Sounds good @ssoloff , it'll be a few days if not couple weeks before I really start digging into DB updates.

**Comment by RoiEXLab** [comment]
@DanVanAtta @ssoloff I actually have a text file lying around on my desktop somewhere listing up a couple of potentially useful changes for the DB.
I'll post it here this evening

**Comment by prastle** [comment]
ty @ all for finally fixing this.  Hopefully all mods will have a lot less headaches of being required to post etc. now.

**Comment by RoiEXLab** [comment]
@DanVanAtta @ssoloff 
Here's the file content:
```
Unique case-insensitive username constraint
unique case-insensitive email constraint
merge banned_* and muted_* into punished_* table
make username bans&mutes username specific
Add banning reason and ban constraint
Add unique ids for users
```

**Comment by DanVanAtta** [comment]
```
merge banned_* and muted_* into punished_* table
make username bans&mutes username specific
```
These could be a bit tricky, i'd be interested in more details. +1 completely agree on the others.

**Comment by RoiEXLab** [comment]
@DanVanAtta 
> merge banned_* and muted_* into punished_* table

If you observe the scheme of the muted_usernames and banned_usernames as well as muted_macs and banned_macs tables, you'll notice absolutely 0 difference except for the table name.
So the idea would be to merge the 4 tables into 2 tables, with an additional column specifying what type of punishment we're talking about.
At this point we can debate if it makes sense to have people muted and banned at the same time, or if we should only apply a unique constraint to the combination of a primary key and the punishment type row, to allow users to be muted and banned at the same time.

> make username bans&mutes username specific

Not even sure on that one, has been some time since I wrote this file. I believe I thought about the way username bans/mutes work: There are 0 checks if the username actually exists, so it's currently possible to ban a user that doesn't exists yet.
While this could make sense to ban certain names with political background, this should be the job of the blacklisted words table (I forgot what it's actually called), which is much more efficient, because the engine checks if the new username contains a banned word, and not if it's the same word.
This change would require a breaking addition of a new network call, to check if the username exists, before it's is sent to the server to get banned. Of course we could simply do server side checks and swallow the error to be backwards compatible instead of throwing an exception, but this really isn't a good practice.
This change would of course also enable us to add constraints to the corresponding tables.

**Comment by prastle** [comment]
off topic but it is also possible atm to ban a typo name, in a bot,  when the name isn't there etc..

**Comment by DanVanAtta** [comment]
@RoiEXLab okay on table merge, makes sense. Table name is a dimension in of itself, when combined we just need to add a new column.

For banning users - for that I think we should start creating identifies of users based on their mac. Then the ban table would use a FK column to indicate which user is banned/muted.

I would otherwise agree to use the blacklisted words table to validate usernames that have been chosen.
